cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -133,10 +133,12 @@ func getDNSIP(client clientset.Interface) (net.IP, error) {
 		return nil, fmt.Errorf("couldn't fetch a valid clusterIP from the kubernetes service")
 	}
 
-	// Build an IP by taking the kubernetes service's clusterIP and appending a "0" and checking that it's valid
-	dnsIP := net.ParseIP(fmt.Sprintf("%s0", k8ssvc.Spec.ClusterIP))
-	if dnsIP == nil {
-		return nil, fmt.Errorf("could not parse dns ip %q: %v", dnsIP, err)
+	// Build an IP by selecting the ninth IP from kubernetes service clusterIP as the DNS IP
+	svcIP := net.ParseIP(k8ssvc.Spec.ClusterIP)
+	if svcIP == nil {
+		return nil, fmt.Errorf("could not parse kubernetes service clutserIP to get the dnsIP %q: %v", svcIP, err)
 	}
+	dnsIP := svcIP.To4()
+	dnsIP[3] += 9
```

Comment:
Why adding `9`?

Reply:
Just to stick with dnsIP which is being assigned today.

Assuming that svcIP is 10.96.0.1 with the above logic the user will get the dnsIP as 10.96.0.10. This is the same IP which a k8s user is getting today. Open for any other suggestion

Reply:
@madhukar32 Well, I think this will result in a conflict. We cannot guarantee `10.96.0.10` has been reserved only for `dnsIP`. It could be a `svcIP`, right?

Reply:
@dixudx: Ok I am new to k8s so I might be completely wrong, please correct me. 

We create this dnsIP in the above code(which is the svcIP for kube-dns svc) and then go and create the actual service [here](https://github.com/kubernetes/kubernetes/blob/fa916c1002992ab0e7e6e044c68b0e52c4ef0a50/cmd/kubeadm/app/phases/addons/dns/dns.go#L75). After kube-dns svc is created it will be actually reserved for kube-dns svc only. Until and unless it is forcefully deleted. So we are pretty much reserving an IP today also isn't it?

Reply:
@madhukar32 Also this value will exceed `255`, right?  What's more, `dnsIP` consumes ip resources from service ip pool, which means we can only create half of services as before. This may cause big trouble for large cluster upgrading.

I think we'd better avoid such hard-coded way.

Reply:
@dixudx: On the value, I have cross verified that it will not exceed 255 <https://play.golang.org/p/KlUAsNrzFJhttps://play.golang.org/p/KlUAsNrzFJ>
Just to clarify here, when I say dnsIP its the service clusterIP of kube-dns svc and it is suppose to be from service ip pool.

Reply:
The 9th ip will only have the same behaviour when the service ip is .1 (but is most consistent this solution than the one we have now)

Reply:
please use `"k8s.io/kubernetes/pkg/registry/core/service/ipallocator".GetIndexedIP(svcSubnet, 10)` or similar for this

Reply:
@luxas: Thanks! Thats a good idea, will update the commit